File: easy_ecom/api/schemas/reports.py
from pydantic import BaseModel
from typing import List, Optional


class ReportDeferredMetric(BaseModel):
    metric: str
    reason: str


class ReportTrendPoint(BaseModel):
    period: str
    value: float


class SalesReport(BaseModel):
    from_date: str
    to_date: str
    sales_count: int
    revenue_total: float
    sales_trend: List[ReportTrendPoint]
    top_products: List[dict]  # { product_id: str, product_name: str, qty_sold: int, revenue: float }
    top_customers: List[dict]  # { customer_id: str, customer_name: str, sales_count: int, revenue: float }
    deferred_metrics: List[ReportDeferredMetric]


class InventoryReport(BaseModel):
    from_date: str
    to_date: str
    total_skus_with_stock: int
    total_stock_units: int
    low_stock_items: List[dict]  # { product_id: str, product_name: str, current_qty: int }
    stock_movement_trend: List[dict]  # { period: str, qty_in: int, qty_out: int }
    inventory_value: Optional[float]
    deferred_metrics: List[ReportDeferredMetric]


class ProductsReport(BaseModel):
    from_date: str
    to_date: str
    highest_selling: List[dict]  # { product_id: str, product_name: str, qty_sold: int, revenue: float }
    low_or_zero_movement: List[dict]  # { product_id: str, product_name: str, qty_sold: int, revenue: float }
    deferred_metrics: List[ReportDeferredMetric]


class FinanceReport(BaseModel):
    from_date: str
    to_date: str
    expense_total: float
    expense_trend: List[dict]  # { period: str, amount: float }
    receivables_total: float
    payables_total: float
    net_operating_snapshot: Optional[float]
    deferred_metrics: List[ReportDeferredMetric]


class ReturnsReport(BaseModel):
    from_date: str
    to_date: str
    returns_count: int
    return_qty_total: int
    return_amount_total: float
    deferred_metrics: List[ReportDeferredMetric]


class PurchasesReport(BaseModel):
    from_date: str
    to_date: str
    purchases_count: int
    purchases_subtotal: float
    purchases_trend: List[dict]  # { period: str, subtotal: float, quantity: int }
    deferred_metrics: List[ReportDeferredMetric]


class ReportsOverview(BaseModel):
    from_date: str
    to_date: str
    sales_revenue_total: float
    sales_count: int
    expense_total: float
    returns_total: int
    purchases_total: float

File: frontend/types/reports.ts
export type ReportDeferredMetric = {
  metric: string;
  reason: string;
};

export type ReportTrendPoint = {
  period: string;
  value: number;
};

export type SalesReport = {
  from_date: string;
  to_date: string;
  sales_count: number;
  revenue_total: number;
  sales_trend: ReportTrendPoint[];
  top_products: Array<{ product_id: string; product_name: string; qty_sold: number; revenue: number }>;
  top_customers: Array<{ customer_id: string; customer_name: string; sales_count: number; revenue: number }>;
  deferred_metrics: ReportDeferredMetric[];
};

export type InventoryReport = {
  from_date: string;
  to_date: string;
  total_skus_with_stock: number;
  total_stock_units: number;
  low_stock_items: Array<{ product_id: string; product_name: string; current_qty: number }>;
  stock_movement_trend: Array<{ period: string; qty_in: number; qty_out: number }>;
  inventory_value: number | null;
  deferred_metrics: ReportDeferredMetric[];
};

export type ProductsReport = {
  from_date: string;
  to_date: string;
  highest_selling: Array<{ product_id: string; product_name: string; qty_sold: number; revenue: number }>;
  low_or_zero_movement: Array<{ product_id: string; product_name: string; qty_sold: number; revenue: number }>;
  deferred_metrics: ReportDeferredMetric[];
};

export type FinanceReport = {
  from_date: string;
  to_date: string;
  expense_total: number;
  expense_trend: Array<{ period: string; amount: number }>;
  receivables_total: number;
  payables_total: number | null;
  net_operating_snapshot: number | null;
  deferred_metrics: ReportDeferredMetric[];
};

export type ReturnsReport = {
  from_date: string;
  to_date: string;
  returns_count: number;
  return_qty_total: number;
  return_amount_total: number;
  deferred_metrics: ReportDeferredMetric[];
};

export type ReportsOverview = {
  from_date: string;
  to_date: string;
  sales_revenue_total: number;
  sales_count: number;
  expense_total: number;
  returns_total: number;
};
